python/lambdas/copy-files-from-tdr/lambda_function.py

```diff
@@ -0,0 +1,84 @@
+import os
+import json
+import boto3
+
+s3_client = boto3.client('s3')
+sqs_client = boto3.client('sqs')
+
+
+def lambda_handler(event, context):
+    destination_bucket = os.environ["DESTINATION_BUCKET"]
+    destination_queue = os.environ["DESTINATION_QUEUE"]
+    for record in event['Records']:
+        body = json.loads(record['body'])
+        file_id = body['fileId']
+        source_bucket = body['bucket']
+        file_location = copy_object(destination_bucket, file_id, source_bucket)
+        copy_object(destination_bucket, f"{file_id}.metadata", source_bucket)
+
+        sqs_client.send_message(QueueUrl=destination_queue, MessageBody=json.dumps({'location': file_location}))
+
+
+def copy_object(destination_bucket, s3_key, source_bucket):
+    try:
+        response = s3_client.head_object(Bucket=source_bucket, Key=s3_key)
+        file_size = response['ContentLength']
+
+        if file_size > 5 * 1024 * 1024 * 1024:
+            return multipart_copy(source_bucket, destination_bucket, s3_key, file_size)
+        else:
+            return standard_copy(source_bucket, destination_bucket, s3_key)
+
+    except Exception as e:
+        print(f"Error processing file {s3_key} from bucket {source_bucket}: {str(e)}")
+        raise e
+
+
+def standard_copy(source_bucket, destination_bucket, s3_key):
+    try:
+        copy_source = {'Bucket': source_bucket, 'Key': s3_key}
+        s3_client.copy_object(CopySource=copy_source, Bucket=destination_bucket, Key=s3_key)
+        print(f"Standard copy of {s3_key} completed successfully.")
+        return f"s3://{destination_bucket}/{s3_key}"
+    except Exception as e:
+        print(f"Error during standard copy of {s3_key}: {str(e)}")
+        raise e
+
+
+def multipart_copy(source_bucket, destination_bucket, s3_key, file_size):
+    try:
+        copy_source = {'Bucket': source_bucket, 'Key': s3_key}
+        multipart_upload = s3_client.create_multipart_upload(Bucket=destination_bucket, Key=s3_key)
+        upload_id = multipart_upload['UploadId']
```

Comment:
If an exception is thrown before this line, then the call on line 83 won't be able to access `upload_id`, which would throw another exception. Maybe we should first declare `upload_id` at the top of the function first?

Reply:
That's what PyCharm was complaining about. I've moved it. It might fail on `create_multipart_upload` but we'll have to risk it.